File: 2020/0010-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        # 如果p为空，返回s是否为空即可
        if not p:
            return not s
        # 当前s、p的第一个值是否匹配
        cur_match = False
        if s:
            cur_match = p[0] in {'.', s[0]}
        # p中有＊号的情况
        if len(p) >= 2 and p[1] == '*':
            # 0个或者多个s[0]字符
            # 比如：(abc, b*abc) or (aabc, a*bc)
            return self.isMatch(s, p[2:]) or (cur_match and self.isMatch(s[1:], p))
        # 最普通的情况
        # 比如：（abc, abc) 或者 (abc, .bc)
        return cur_match and self.isMatch(s[1:], p[1:])
            

File: hot100/0010-regular-expression-matching.go
func isMatch(s string, p string) bool {
    // 考虑s或者p为空的情况
    if p == "" {
        return s == ""
    }

    // 先考虑第一个字符的匹配情况
    var match bool
    if len(s) > 0 && (p[0] == '.' || s[0] == p[0]) {
        match = true
    }

    // 考虑两种特殊的带*情况
    // 1）*号代表了0次重复前一字符，比如：s="abc", p="b*abc"
    // 2）*号代表了1+次重复前一字符，比如：s="bbabc", p="b*abc" 或者 p=".*abc"
    if len(p) >= 2 && p[1] == '*' {
        return isMatch(s, p[2:]) || (match && isMatch(s[1:], p))
    }

    // 普通情况
    return match && isMatch(s[1:], p[1:])
}
